File: 3002-maximum-size-of-a-set-after-removals/3002-maximum-size-of-a-set-after-removals.py
class Solution:
    def maximumSetSize(self, nums1: List[int], nums2: List[int]) -> int:
        s1, s2 = set(nums1), set(nums2)
        n, n1, n2, c = len(nums1), len(s1), len(s2), len(s1 & s2)
        return min(n, min(n1 - c, n // 2) + min(n2 - c, n // 2) + c)

File: 3002-maximum-size-of-a-set-after-removals/3002-maximum-size-of-a-set-after-removals.java
class Solution {
    public int maximumSetSize(int[] nums1, int[] nums2) {
        Set<Integer> s1 = new HashSet(); 
        Set<Integer> s2 = new HashSet();
        Set<Integer> s3 = new HashSet(); 
        
        for (int num: nums1){
            s1.add(num);
            s3.add(num);
        } 
        for (int num: nums2) {
            s2.add(num);
            s3.add(num);
        } 
          
        return Math.min(Math.min(s1.size(), nums1.length/2) + 
                        Math.min(s2.size(), nums2.length/2), s3.size());                            
    }
}